src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs

```diff
@@ -77,27 +78,34 @@ private PackageManagerControl()
             InitializeComponent();
         }
 
-        public static async ValueTask<PackageManagerControl> CreateAsync(PackageManagerModel model, INuGetUILogger uiLogger)
+        public static async ValueTask<PackageManagerControl> CreateAsync(PackageManagerModel model, INuGetUILogger uiLogger, CancellationToken cancellationToken)
         {
             Assumes.NotNull(model);
 
             var packageManagerControl = new PackageManagerControl();
-            await packageManagerControl.InitializeAsync(model, uiLogger);
+            await packageManagerControl.InitializeAsync(model, uiLogger, cancellationToken);
             return packageManagerControl;
         }
 
-        private async ValueTask InitializeAsync(PackageManagerModel model, INuGetUILogger uiLogger)
+        private async ValueTask InitializeAsync(PackageManagerModel model, INuGetUILogger uiLogger, CancellationToken cancellationToken)
         {
-            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
             _sinceLastRefresh = Stopwatch.StartNew();
 
             _installedTabTelemetryData = new PackageManagerInstalledTabData();
 
             Model = model;
             _uiLogger = uiLogger;
-            Settings = await ServiceLocator.GetComponentModelServiceAsync<ISettings>();
 
+            await TaskScheduler.Default;
```

Comment:
Does CreateDocWindowForSolutionAsync ends up waiting on this right? 

I think we might have introduced a race here; previously CreateDocWindowForSolutionAsync assumed that FindExistingSolutionWindowFrameAsync had already checked if the window was open but that check is only relevant while you are on the UI thread (someone can open the window behind your back when you get off UI thread).

I presume previously despite being marked as async, these were mostly stayed on the UI thread and remain synchronous. Now, we're actually hitting real "yields" there's a race; if you invoke the method twice in succession very quickly, this path I think will be hit twice and the window will get opened twice or whatever happens in that case.

You can simulate what this be like on slow machines or high thread pool latency by putting a Task.Delay(15000) here, and then invoke the method command twice in a row; what happens?

Other VS tool windows don't have this problem, because they say on the UI thread between the check and the creation of the window, or use the built-in async tool window logic for asynchronously loading the tool window.

Reply:
I thought I responded to this. I did the following:
1. Set a 15000 delay after the line commented on. 
2. Started debugging
3. Right clicked solution and opened PM UI
4. Repeated step 3
5. Waited

When it completed, I only had one PM UI window open and it was working fine. I tried the same thing for the project level PM UI. 

I didn't get any exceptions in the debugger either. When I stepped through the debugger it wasn't obvious to me why a new window wasn't created.